core/src/main/java/jenkins/install/UpgradeWizard.java

```diff
@@ -0,0 +1,147 @@
+package jenkins.install;
+
+import hudson.Extension;
+import hudson.model.PageDecorator;
+import hudson.model.UpdateSite.Plugin;
+import hudson.util.HttpResponses;
+import hudson.util.VersionNumber;
+import jenkins.model.Jenkins;
+import org.apache.commons.io.FileUtils;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.HttpResponse;
+import org.kohsuke.stapler.interceptor.RequirePOST;
+
+import javax.inject.Inject;
+import java.io.File;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Logger;
+
+import static java.util.logging.Level.*;
+import static org.apache.commons.io.FileUtils.*;
+import static org.apache.commons.lang.StringUtils.*;
+
+/**
+ * This is a stop-gap measure until JENKINS-33663 comes in.
+ * This call may go away. Please don't use it from outside.
+ *
+ * @author Kohsuke Kawaguchi
+ */
+@Restricted(NoExternalUse.class)
+@Extension
+public class UpgradeWizard extends PageDecorator {
+    @Inject
+    private Jenkins jenkins;
+
+    /**
+     * Is this instance fully upgraded?
+     */
+    private volatile boolean upToDate;
+
+    /**
+     * File that captures the state of upgrade.
+     *
+     * This file records the vesrion number that the installation has upgraded to.
+     */
+    /*package*/ File getStateFile() {
+        return new File(Jenkins.getInstance().getRootDir(),"upgraded");
```

Comment:
Valid point, but I also need to get this change badly in before beta.

I'll come back to this tonight, and if I can get that done I will. Otherwise let's merge this and flag this in the next step that @kzantow is working on, so that we can fix this in there.